service_management/app/services/analysis/SampleObjectService.scala

```diff
@@ -96,4 +93,29 @@ class SampleObjectService @Inject()(
     soDao.listForMuseum(mid)
   }
 
+  def delete(
+      oid: ObjectUUID
+  )(implicit currUser: AuthenticatedUser): Future[MusitResult[Unit]] = {
+    def enrich(orig: SampleObject) = {
+      orig.copy(
+        isDeleted = true,
+        updatedStamp = Some(ActorStamp(currUser.id, dateTimeNow))
+      )
+    }
+
+    val updatedRes = for {
+      orig <- MusitResultT(findById(oid, MusitEmpty))
+      _    <- MusitResultT(soDao.update(enrich(orig)))
+      upd <- MusitResultT(
+              findById(
+                oid,
+                MusitInternalError(s"Couldn't find sample $oid after delete")
+              )
+            )
+    } yield upd
+
+    // Need to do some tricks to align the shapes again.
+    updatedRes.value.map(_.map(_ => MusitSuccess(())))
```

Comment:
The function finally maps into the `Future` and ignores the updated `SampleObject` value completely. Which is perfectly reasonable since it's been marked as deleted.

This tells us that the code might be doing more than it needs to. The unnecessary step in this case is Step 4.

Let's see how we can make the for-comprehension block a bit simpler:

```scala
def delete(... args ...) {
  ...
  val updatedRes = for {
    orig <- MusitResultT(findById(oid, MusitEmpty))
    del  <- MusitResultT(soDao.update(enrich(orig)).map(_ => MusitSuccess(()))
  } yield del

  updatedRes.value
}
```

Step 4 is now gone, as well as the last `map` function to transform `updatedRes`.